Applications/Spire/Source/Ui/RealSpinBox.cpp

```diff
@@ -0,0 +1,441 @@
+  m_step = step;
+}
+
+
+RealSpinBox::Real RealSpinBox::get_value() const {
+  return m_last_valid_value;
+}
+
+void RealSpinBox::set_value(Real value) {
+  blockSignals(true);
+  value = clamped_real(value, m_minimum, m_maximum);
+  lineEdit()->setText(display_string(value));
+  blockSignals(false);
+}
+
+void RealSpinBox::stepBy(int step) {
+  add_step(step);
+}
+
+connection RealSpinBox::connect_change_signal(
+    const ChangeSignal::slot_type& slot) const {
+  return m_change_signal.connect(slot);
+}
+
+void RealSpinBox::add_step(int step) {
+  add_step(step, Qt::NoModifier);
+}
+
+void RealSpinBox::add_step(int step, Qt::KeyboardModifiers modifiers) {
+  step = [&] {
+    if(modifiers == Qt::ShiftModifier) {
+      return step * SHIFT_STEPS;
+    }
+    return step;
+  }();
+  auto value = get_value(text());
+  if(text().isEmpty() || value == boost::none) {
+    auto value = Real(step);
+    value *= m_step;
+    lineEdit()->setText(display_string(value));
+    lineEdit()->setCursorPosition(text().length());
+    return;
+  }
+  *value += step;
+  value = clamped_real(*value, m_minimum, m_maximum);
+  lineEdit()->setText(display_string(*value));
+  lineEdit()->setCursorPosition(text().length());
+}
+
+QString RealSpinBox::display_string(Real value) {
+  auto str = text();
+  str.remove(QRegularExpression("0+$"));
+  str.remove(QRegularExpression(
+    QString("\\%1$").arg(m_locale.decimalPoint())));
+  if(str.contains(m_locale.decimalPoint())) {
+    auto displayed_text = QString::fromStdString(value.str(
+      str.length() - str.indexOf(m_locale.decimalPoint()) - 1,
+      std::ios_base::fixed));
+    return displayed_text.remove(
+      QRegularExpression(QString("\\%1%2*$").arg(m_locale.decimalPoint())
+      .arg(0))).trimmed();
+  }
+  return QString::fromStdString(value.str(text().length() + 1,
+    std::ios_base::dec));
+}
+
+QStyle::SubControl RealSpinBox::get_current_control(const QPoint& mouse_pos) {
+  QStyleOptionSpinBox opt;
+  initStyleOption(&opt);
+  opt.subControls = QStyle::SC_All;
+  return style()->hitTestComplexControl(QStyle::CC_SpinBox, &opt, mouse_pos,
+    this);
+}
+
+boost::optional<RealSpinBox::Real> RealSpinBox::get_value(
+    const QString& text) const {
+  try {
+    return Real(text.toStdString().c_str());
+  } catch (const std::runtime_error&) {
+    return boost::none;
+  }
+}
+
+bool RealSpinBox::is_valid(const QString& text) {
+  if(!text.contains(m_real_regex)) {
+    return false;
+  }
+  if(!text.contains(QRegularExpression("[0-9]"))) {
+    return true;
+  }
+  auto value = get_value(text);
+  if(value == boost::none) {
+    return true;
+  }
+  return value->compare(m_minimum) >= 0 && value->compare(m_maximum) <= 0;
+}
+
+void RealSpinBox::set_stylesheet(bool is_up_disabled, bool is_down_disabled) {
+  if(isEnabled()) {
+    setButtonSymbols(UpDownArrows);
+  } else {
+    setButtonSymbols(NoButtons);
+  }
+  auto [up_icon, up_icon_hover] = [&] () -> std::pair<QString, QString> {
```

Comment:
Use ```std::tuple```, remove the ```-> ...```

    auto [up_icon, up_icon_hover] = [&] {
      if(is_up_disabled) {
        return std::tuple(...);
      }
      return std::tuple(...);
    }();